gk/fib.c

```diff
@@ -1664,3 +1666,300 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
+	static uint32_t correct_ctypeid_fib_dump_entry;
+
+	struct gk_fib *fib;
+	const struct rte_lpm_rule *re4;
+	struct rte_lpm_iterator_state state;
+	struct gk_fib_dump_entry dentry;
+	void *cdata;
+
+	if (!assigned_type_fib_dump_entry) {
+		correct_ctypeid_fib_dump_entry = luaL_get_ctypeid(l,
+			CTYPE_STRUCT_FIB_DUMP_ENTRY_PTR);
+		assigned_type_fib_dump_entry = true;
+	}
+
+	ret = rte_lpm_iterator_state_init(ltbl->lpm, 0, 0, &state);
+	if (ret < 0)
+		luaL_error(l, "gk: failed to initialize the lpm rule iterator state at %s!",
+			__func__);
+
+	index = rte_lpm_rule_iterate(&state, &re4);
+	while (index >= 0) {
+		fib = &ltbl->fib_tbl[re4->next_hop];
+		if (fib->action == GK_FWD_NEIGHBOR_FRONT_NET ||
+				fib->action == GK_FWD_NEIGHBOR_BACK_NET) {
+			index = rte_lpm_rule_iterate(&state, &re4);
+			continue;
+		}
+
+		memset(&dentry, 0, sizeof(dentry));
+		dentry.addr.proto = ETHER_TYPE_IPv4;
+		dentry.addr.ip.v4.s_addr = htonl(re4->ip);
+		dentry.prefix_len = state.depth;
+		fillup_gk_fib_dump_entry(&dentry, fib);
+
+		lua_pushvalue(l, 2);
+		lua_insert(l, 3);
+		cdata = luaL_pushcdata(l, correct_ctypeid_fib_dump_entry,
+			sizeof(struct gk_fib_dump_entry *));
+		*(struct gk_fib_dump_entry **)cdata = &dentry;
+		lua_insert(l, 4);
+
+		if (lua_pcall(l, 2, 1, 0) != 0) {
+			/*
+			 * The following call isn't strictly necessary
+			 * because we are going to raise a Lua error.
+			 */
+			lua_error(l);
+		}
+
+		index = rte_lpm_rule_iterate(&state, &re4);
+	}
+}
+
+static void
+list_ipv6_fib_entries(lua_State *l, struct gk_lpm *ltbl)
+{
+	int ret, index;
+
+	static bool assigned_type_fib_dump_entry = false;
+	static uint32_t correct_ctypeid_fib_dump_entry;
+
+	struct gk_fib *fib;
+	struct rte_lpm6_rule re6;
+	struct rte_lpm6_iterator_state state6;
+	struct gk_fib_dump_entry dentry;
+	void *cdata;
+
+	if (!assigned_type_fib_dump_entry) {
+		correct_ctypeid_fib_dump_entry = luaL_get_ctypeid(l,
+			CTYPE_STRUCT_FIB_DUMP_ENTRY_PTR);
+		assigned_type_fib_dump_entry = true;
+	}
+
+	ret = rte_lpm6_iterator_state_init(ltbl->lpm6, NULL, 0, &state6);
+	if (ret < 0)
+		luaL_error(l, "gk: failed to initialize the lpm6 rule iterator state at %s!",
+			__func__);
+
+	index = rte_lpm6_rule_iterate(&state6, &re6);
+	while (index >= 0) {
+		fib = &ltbl->fib_tbl6[re6.next_hop];
+		if (fib->action == GK_FWD_NEIGHBOR_FRONT_NET ||
+				fib->action == GK_FWD_NEIGHBOR_BACK_NET) {
+			index = rte_lpm6_rule_iterate(&state6, &re6);
+			continue;
+		}
+
+		memset(&dentry, 0, sizeof(dentry));
+		dentry.addr.proto = ETHER_TYPE_IPv6;
+		rte_memcpy(&dentry.addr.ip.v6, re6.ip,
+			sizeof(dentry.addr.ip.v6));
+		dentry.prefix_len = re6.depth;
+		fillup_gk_fib_dump_entry(&dentry, fib);
+
+		lua_pushvalue(l, 2);
+		lua_insert(l, 3);
+		cdata = luaL_pushcdata(l, correct_ctypeid_fib_dump_entry,
+			sizeof(struct gk_fib_dump_entry *));
+		*(struct gk_fib_dump_entry **)cdata = &dentry;
+		lua_insert(l, 4);
+
+		if (lua_pcall(l, 2, 1, 0) != 0) {
```

Comment:
Since we are just raising the Lua error further here, we can drop the `if` statement and just rename `lua_pcall()` with `lua_call()`.